Friboo 2.0.0

dryewo opened the pull request:
The New HTTP component is decomplected from Zalando-specific details and made highly configurable. Currently uses `org.zalando.stups.friboo.system.new-http`, but then to be moved into `org.zalando.stups.friboo.system.http`, replacing the old one and introducing a breaking change, hence 2.0.0-SNAPSHOT.

Addresses #53, #51, #49, #44, #113 

Work in Progress, opened for early feedback.
Still lacks tests.

UPD. Tests are written.
Still need to isolate all zalando-specific details in a separate namespace.

UPD. Dependencies are isolated.
Need to try out migrating some existing microservices to Friboo 2.0 and eventually write a migration guide.

UPD. [Essentials](https://github.com/zalando-stups/essentials) now runs on friboo-ext-zalando 2.0.0-beta1 (which is based on friboo 2.0.0-beta1) in production.

UPD. README is updated to reflect 2.0.0

Commit:
Introduce New HTTP component.
Version 2.0.0-SNAPSHOT.

Commit:
Remove remaining http2 mentions.

danpersa commented on test/org/zalando/stups/friboo/system/new_http_test.clj:
```diff
@@ -0,0 +1,6 @@
+            [midje.sweet :refer :all]
+            [org.zalando.stups.friboo.system.http :refer :all]))
+
+;; TODO Write tests here
```
you should either write the tests either delete this file ;)

danpersa commented on src/org/zalando/stups/friboo/system/new_http.clj:
```diff
@@ -0,0 +1,275 @@
+  [coll]
+  (apply concat coll))
+
+(defn with-flattened-options-map
```
Shouldn't you make some of the functions private? This would make a better statement on what the public API for this ns is.

dryewo replied:
It's a nice idea actually, as making them private doesn't prevent testing at all (you can always call a function using `#'new-http/with-flattened-options-map`

harti2006 commented on src/org/zalando/stups/friboo/system/new_http.clj:
```diff
@@ -0,0 +1,275 @@
+                    (middleware-chain (:before-parser middlewares))
+                    (with-flattened-options-map s1st/parser (:parser s1st-options))
+
+                    (middleware-chain (:before-protector middlewares))
```
We should really check if the protection step can already happen before the parsing. I don't see why we would need parsed parameters in order to evaluate the Authorization header

prayerslayer commented on src/org/zalando/stups/friboo/system/new_http.clj:
```diff
@@ -0,0 +1,275 @@
+      (clojure.string/split #":" 2)
+      (#(zipmap [:username :password] %))))
+
+;; TODO Zalando specific
```
That and also I think it was added there because it was necessary for `pierone login`?

harti2006 replied:
Exactly.

dryewo replied:
Yes, that's why I suggest to move it into "serious-friboo" later. It definitely should not be the default behavior.

prayerslayer replied:
Since no other of our friboo apps has this behavior, I think, it may as well be moved into pierone.

Commit:
Tests for new-http.
More comments.

Commit:
Merge branch 'master' into new-http-component

Commit:
Extract zalando-specific stuff into a subpackage (not everything yet).

Commit:
Fix protocol import

Commit:
Refactor Authorizer.
More system tests.

Commit:
Remove old http namespace, rename new-http -> http

Commit:
More tests.

Commit:
Merge branch 'master' into new-http-component

Commit:
Refactor DB component, add tests requiring PostgreSQL.

dryewo commented:
@prayerslayer please take a look at the refactored auth.clj and examples in zalando_specific/system_test.clj

PetrGlad commented:
Maybe rename o.z.s.f.zalando-specific -> o.z.s.f.zalando-internal

I think that ideally this should be in a separate project.

PetrGlad commented on src/org/zalando/stups/friboo/system/http.clj:
```diff
@@ -11,41 +11,141 @@
+    (throw (ex-info "Component already started, aborting." {})))
+  ;; Refer to default-middlewares object below for middleware lists
+  ;; Create context from the YAML definition on the classpath, optionally provide validation flag
+  (let [handler (-> (apply s1st/context :yaml-cp api-resource (flatten1 (:context s1st-options)))
```
// If I wrote swagger1st I'd have option maps as maps to not convert them forth and back in every place. In particular, it would streamline this piece substantially.

dryewo replied:
Could be even better, amazonica (AWS SDK wrapper) supports both ways.

PetrGlad replied:
More ways is not always better.

Commit:
Refactor Authorizer (breaking change):
- make parameter order consistent
- pass only tokeninfo
- expect entire payload instead of just team

Commit:
Add make.sh utility script for starting a database for unit tests

Commit:
"unknown user" makes it look ugly

Commit:
Add CHANGELOG.md, describe breaking changes there

Commit:
Adapt leiningen template to Friboo 2.0.0-SNAPSHOT, remove Zalando-specific features.

Commit:
If a component throws an exception during system start, stop all started components to clean up their resources.

Commit:
Merge remote-tracking branch 'origin/master' into new-http-component

# Conflicts:
#	test/org/zalando/stups/friboo/system/db_test.clj

Commit:
Version 2.0.0-beta1-SNAPSHOT

Commit:
Version 2.0.0-beta1

Commit:
Version 2.0.0-beta2-SNAPSHOT

Commit:
Update SCM links in metadata

Commit:
Merge remote-tracking branch 'origin/master' into new-http-component

# Conflicts:
#	project.clj

Commit:
Version 2.0.0-beta2

Commit:
Version 2.0.0-beta3-SNAPSHOT

Commit:
Version 2.0.0-beta3

Commit:
Version 2.0.0-beta4-SNAPSHOT

Commit:
Move zalando-specific parts to friboo-ext-zalando library.
Make metrics prefix configurable instead of just "zmon".

Commit:
Bump dependencies

Commit:
Version 2.0.0-beta4

Commit:
Version 2.0.0-beta5-SNAPSHOT

Commit:
fix travis badge

Commit:
Add a namespace for dev/test utils

Commit:
Update CHANGELOG

Commit:
Add dev.clj with development utils

Commit:
Version 2.0.0-beta5

Commit:
Version 2.0.0-beta6-SNAPSHOT

Commit:
Update leiningen template to use friboo 2.x

Commit:
workarounds to fix template releasing

Commit:
prefix template tags with 'template-'

Commit:
Version 2.0.0-beta1

Commit:
Version 2.0.0-beta2-SNAPSHOT

dryewo commented:
Template released: https://clojars.org/friboo/lein-template/versions/2.0.0-beta1

    $ lein new friboo com.example/foo-bar --template-version 2.0.0-beta1

Commit:
Update README

Commit:
Remove license headers from files.

prayerslayer commented on src/org/zalando/stups/friboo/system/metrics.clj:
```diff
@@ -51,22 +39,22 @@
-            status (:status response)
-            timer (tmr/timer metrics-registry ["zmon" "response" (swagger1st-request2string request status)])]
+            status   (:status response)
+            ;; TODO replace "zmon" with configuration value
```
That's done, I guess?

prayerslayer replied:
Also, do I need to provide it all the time or is there a default value (doesn't look like it)?

dryewo replied:
Yes, it's done.
By default it's "friboo", see `default-configuration` on line 100.
For internal Zalando projects "zmon" will be set: https://github.com/zalando-incubator/friboo-ext-zalando/blob/master/src/org/zalando/stups/friboo/zalando_internal/config.clj#L10

Commit:
Update changelog

Commit:
remove done todo

prayerslayer commented on template/project.clj:
```diff
@@ -1,10 +1,20 @@
+(defproject friboo/lein-template "2.0.0-beta2-SNAPSHOT"
   :description "Leiningen template for Friboo library"
-  :url "https://github.com/dryewo/friboo-template"
+  :url "https://github.com/zalando/friboo/template"
```
404

prayerslayer commented:
👍

dryewo commented:
👍
